### PR TITLE
Enable lombok with javac

### LINKED ISSUE #955: javac doesn't work with lombok
Steps to reproduce:

- start Eclipse with javac and the lombok agent
- create
```
package com.example;
import lombok.AllArgsConstructor;
import lombok.Data;
public class Main {
	public void test() {
		Person p = new Person(10, "abc");
		System.out.println(p.getAge() + p.getName());
		System.out.println(p.getAge() + p.getName());
		System.out.println(p.getAge() + p.getName());
	}
	@Data
	@AllArgsConstructor
	class Person {
		private Integer age;
		private String name;

	}
}
```
you will get seven errors and two warnings

### PR DESCRIPTION
**Pull request by snjeza**
Fixes #955 

Test project: 
[testlombok.zip](https://github.com/user-attachments/files/17938190/testlombok.zip)

### CODE REVIEW

**Comment by mickaelistria** [comment]
We've had demo of lombok working without it, just having it in the classpath used to work. What makes that changes are necessary now?

**Comment by snjeza** [comment]
> We've had demo of lombok working without it, just having it in the classpath used to work. What makes that changes are necessary now?

Could you try to open the attached project?
You will get several errors in Main.java

**Comment by snjeza** [comment]
javac/lombok without this PR:
lombokwithout

with the PR
lombokwith

**Review by datho7561 — APPROVED** [review]
I confirmed this is needed to get lombok working properly.

It seems bad that we need to parse twice to compile. I don't fully understand what's going on, so I don't know if it's possible to avoid this. However, I think it's better to get it working than to have it fast, so I think this change is good as-is.

There is some debug code in here to keep track of how long analysis takes, both for the. Perhaps this should be removed and replaces with a progress monitor subtask?

**Comment by snjeza** [comment]
> It seems bad that we need to parse twice to compile. I don't fully understand what's going on, so I don't know if it's possible to avoid this. However, I think it's better to get it working than to have it fast, so I think this change is good as-is.

You can set a breakpoint at https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/blob/bf00d42221a92807f1f019181e12971363f36c6a/org.eclipse.jdt.core.javac/src/org/eclipse/jdt/core/dom/JavacCompilationUnitResolver.java#L712

`task.parse()` calls `JavacTaskImpl.prepareCompiler(true)` which calls
- https://github.com/openjdk/jdk/blob/f6d29909bb5cce0a73945c9fccc9c3fa7d4804fd/src/jdk.compiler/share/classes/com/sun/tools/javac/api/JavacTaskImpl.java#L213
```
 compiler.initProcessAnnotations(processors, args.getFileObjects(), args.getClassNames());
```

You can also try the following without the PR:

- open Main.java - you will see seven errors
- open Person.java
- activate Main.java - you will still see the errors
- activate Person.java; edit it without saving
- activate Main.java - errors are gone; javac run Annotation Processing on Main.java and Person.java

**Comment by snjeza** [comment]
> There is some debug code in here to keep track of how long analysis takes, both for the. Perhaps this should be removed and replaces with a progress monitor subtask?

I manually set `private static boolean DEBUG = true` to check the performance. I have seen a similar code at https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/blob/bf00d42221a92807f1f019181e12971363f36c6a/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/parser/Parser.java#L88
Do you want me to delete it?

**Comment by snjeza** [comment]
addDependentFileObjects is pretty fast. We skip linting, annotation processing...

```
!ENTRY org.eclipse.jdt.core 1 0 2024-11-27 23:10:51.125
!MESSAGE addDependentFileObjects analyze elements=2, took: 20, first=Main

!ENTRY org.eclipse.jdt.core 1 0 2024-11-27 23:10:51.315
!MESSAGE analyze elements=2, took: 186, first=Main

!ENTRY org.eclipse.jdt.core 1 0 2024-11-27 23:10:56.033
!MESSAGE addDependentFileObjects analyze elements=2, took: 28, first=Main

!ENTRY org.eclipse.jdt.core 1 0 2024-11-27 23:10:56.204
!MESSAGE analyze elements=2, took: 166, first=Main

!ENTRY org.eclipse.jdt.core 1 0 2024-11-27 23:10:57.211
!MESSAGE addDependentFileObjects analyze elements=2, took: 21, first=Main

!ENTRY org.eclipse.jdt.core 1 0 2024-11-27 23:10:57.379
!MESSAGE analyze elements=2, took: 164, first=Main

!ENTRY org.eclipse.jdt.core 1 0 2024-11-27 23:10:59.979
!MESSAGE addDependentFileObjects analyze elements=2, took: 20, first=Main

!ENTRY org.eclipse.jdt.core 1 0 2024-11-27 23:11:00.154
!MESSAGE analyze elements=2, took: 171, first=Main
```

**Comment by datho7561** [comment]
> I have seen a similar code at ... Do you want me to delete it?

I think we should remove that code. If someone wants to debug how long these analysis steps take, they can always add it back, but for now it will make the code cluttered.

> addDependentFileObjects is pretty fast

Okay, my main concern about running analysis twice was performance, so if it's fast this isn't an issue.

**Comment by mickaelistria** [comment]
I will try to investigate it later, but really none of this is supposed to be necessary: the Javac specification and documentation, together with our previous experiences, do show that just having lombok in the classpath should be enough and so we should just use it expected without further workarounds. Maybe the issue is that we force `-proc:none`?

**Comment by datho7561** [comment]
@mickaelistria what was the hack that ensured that only the classes passed in to javac were processed, was that `-proc:none`?

**Comment by snjeza** [comment]
> I think we should remove that code. If someone wants to debug how long these analysis steps take, they can always add it back, but for now it will make the code cluttered.

I have removed it.

**Comment by snjeza** [comment]
>  what was the hack that ensured that only the classes passed in to javac were processed, was that -proc:none?

We can't control how javac do annotation processing. See https://github.com/openjdk/jdk/blob/f6d29909bb5cce0a73945c9fccc9c3fa7d4804fd/src/jdk.compiler/share/classes/com/sun/tools/javac/api/JavacTaskImpl.java#L213

`-proc:none` just skips annotation processing.

**Comment by snjeza** [comment]
> Maybe the issue is that we force -proc:none?

Right, that is the issue when building/compiling a project. I have fixed it by https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/pull/992/commits/8b11f68aa64fc692196466d9cb10afb4f5a8bfa1#diff-b5085216705bfeedc7f377a098787482377d9d4a4bf8915c8db03bb0e587f5bdR184
We don't have the issue with dependent classes here because javac builder/compiler compiles the whole project.

The second issue is related to reconciling working units. We add the current working copy and dirty working copies to javac. javac properly compiles them and finds dependent classes, but it skips annotation processing on dependent classes.

> the Javac specification and documentation, together with our previous experiences, do show that just having lombok in the classpath should be enough and so we should just use it expected without further workarounds.

This works when compiling the whole project. I will try to check how NetBeans and IntelliJ IDEA do it.

**Comment by snjeza** [comment]
@datho7561 @mickaelistria  I have updated the PR.
The PR fixes the issue in the following way:
- enables annotation processing when building/compiling a project - https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/pull/992/commits/3ed24e9361aa0bcf555aa1acee7d5157a79db74c#diff-b5085216705bfeedc7f377a098787482377d9d4a4bf8915c8db03bb0e587f5bdR184
- add the project's output directory to the classpath when reconciling working copy/copies - https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/pull/992/commits/3ed24e9361aa0bcf555aa1acee7d5157a79db74c#diff-bd7a44684bdcb41a3509ee1d2a4194d53f5baa9ae6663eb6c20978f773849b7dR362

**Review by datho7561 — APPROVED** [review]

File: org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacCompiler.java [hunk 1]
```diff
@@ -179,6 +179,9 @@ public int errorCount() {
 					return this.isInGeneration ? 0 : super.errorCount();
 				}
 			};
+			// https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/issues/955
+			// enable annotation processing
+			javacOptions.put(Option.A, "");
```
Comment (mickaelistria) [inline org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacCompiler.java]: This is more or less already done in JavacUtils lines 183..186. However, JDT mandates that annotation processing is enabled to consider it. Maybe the issue is just that the necessary JDT flags were missing?

File: org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacUtils.java [hunk 2]
```diff
@@ -357,6 +357,10 @@ public static <T> boolean isEmpty(List<T> list) {
 	private static Collection<File> classpathEntriesToFiles(JavaProject project, Predicate<IClasspathEntry> select) {
 		try {
 			LinkedHashSet<File> res = new LinkedHashSet<>();
+			IPath outputLocation = project.getOutputLocation();
+			if (outputLocation != null) {
+				addPath(project, outputLocation, res);
+			}
```
Comment (mickaelistria) [inline org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacUtils.java]: Similar operation is already performed on line 307 with `classpathFiles.addAll(outDirectories(javaProject, entry -> isTest || !entry.isTest()));`.

Comment (snjeza) [reply org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacUtils.java]: This code is called only for a modular project. I have updated the PR.

File: org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacCompiler.java (hunk 1 above)
Comment (snjeza) [reply org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacCompiler.java]: You are right. I have fixed JavacUtils lines 183..186.

**Comment by snjeza** [comment]
@mickaelistria @datho7561 I have updated the PR.
Now, you have to enable annotation processing.
~/.settings/org.eclipse.jdt.apt.core.prefs
```
eclipse.preferences.version=1
org.eclipse.jdt.apt.aptEnabled=true
org.eclipse.jdt.apt.genSrcDir=.apt_generated
org.eclipse.jdt.apt.genTestSrcDir=.apt_generated_tests
org.eclipse.jdt.apt.reconcileEnabled=true
```

File: org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacUtils.java [hunk 3]
```diff
@@ -182,7 +182,10 @@ private static void configureOptions(IJavaProject javaProject, Context context,
 
 		Map<String, String> processorOptions = ProcessorConfig.getProcessorOptions(javaProject);
 		for (Entry<String, String> processorOption : processorOptions.entrySet()) {
-			options.put("-A" + processorOption.getKey() + "=" + processorOption.getValue(), Boolean.toString(true));
+			options.put("-A", processorOption.getKey() + "=" + processorOption.getValue());
```
Comment (mickaelistria) [inline org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacUtils.java]: Can we also use `Options.A` here by the way?

File: org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacUtils.java [hunk 4]
```diff
@@ -354,6 +357,12 @@ public static <T> boolean isEmpty(List<T> list) {
 	private static Collection<File> classpathEntriesToFiles(JavaProject project, Predicate<IClasspathEntry> select) {
 		try {
 			LinkedHashSet<File> res = new LinkedHashSet<>();
+			if (project.getModuleDescription() == null) {
+				IPath outputLocation = project.getOutputLocation();
+				if (outputLocation != null) {
+					addPath(project, outputLocation, res);
+				}
+			}
```
Comment (mickaelistria) [inline org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacUtils.java]: Adding the project output to classpath should already be done with
```
else if (select.test(current)) {
	IPath path = current.getPath();
	addPath(project, path, res);
}
```
around line 388.
So I don't think it's necessary as part of this change.

**Review by mickaelistria — COMMENTED** [review]
Please see comments inline.
As Lombok seems rather fragile, do you think you can create a basic test case from this example?

File: org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacUtils.java (hunk 3 above)
Comment (snjeza) [reply org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacUtils.java]: Options.A only enables annotation processing. We also need to add -Akey=value to all annotation processors.

**Comment by snjeza** [comment]
> Please see comments inline.
As Lombok seems rather fragile, do you think you can create a basic test case from this example?

~~I will try.~~
I have added a test.

File: org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacUtils.java (hunk 3 above)
Comment (mickaelistria) [reply org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacUtils.java]: Option.A is just a richer definition for `-A`. Looking at how Option.A is consumed, I think it should be good to use it here too with `options.put(Option.A, processorOption.getKey() + '=' + processorOptions.getValue())`.

File: org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacUtils.java (hunk 4 above)
Comment (snjeza) [reply org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacUtils.java]: > Adding the project output to classpath should already be done with

This code adds the source entry but not the output location.
You can try 
[testlombok.zip](https://github.com/user-attachments/files/17980112/testlombok.zip)

File: org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacUtils.java (hunk 3 above)
Comment (snjeza) [reply org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacUtils.java]: Fixed.

File: org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacUtils.java (hunk 4 above)
Comment (mickaelistria) [reply org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacUtils.java]: > This code adds the source entry but not the output location.

Oops, I thought I had fixed that; thanks for spotting it.
Can you please submit it as a separate PR as it's desired independently for lombok?

Comment (snjeza) [reply org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacUtils.java]: I have created https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/pull/1006 and updated this PR.

Comment (snjeza) [reply org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacUtils.java]: #1006 has been merged. I have rebased this PR to include #1006.

File: org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacUtils.java (hunk 3 above)
Comment (testforstephen) [reply org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacUtils.java]: Have you tested it with `options.put(Option.A, processorOption.getKey() + '=' + processorOptions.getValue())`? From my last experiment, `options.put("-A" + processorOption.getKey() + "=" + processorOption.getValue(), Boolean.toString(true));` works well.

Comment (snjeza) [reply org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacUtils.java]: You can try the following:
- add `org.eclipse.jdt.apt.processorOptions/a=b` to ~/.settings/org.eclipse.jdt.apt.core.prefs
- apply the following patch
```
diff --git a/org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacUtils.java b/org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacUtils.java
index d52462a5b8..536879c434 100644
--- a/org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacUtils.java
+++ b/org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacUtils.java
@@ -184,6 +184,9 @@ public class JavacUtils {
                Map<String, String> processorOptions = ProcessorConfig.getProcessorOptions(javaProject);
                for (Entry<String, String> processorOption : processorOptions.entrySet()) {
                        options.put(Option.A, processorOption.getKey() + "=" + processorOption.getValue());
+                       if (!options.isSet(Option.A)) {
+                               ILog.get().error("Option.A is not set");
+                       }
                }
                if (!options.isSet(Option.A) && ProcessorConfig.isAnnotationProcessingEnabled(javaProject)) {
                        options.put(Option.A, "");
```
- recompile the project (Project>Clean)
- check the error log. You should not see `Option.A is not set`

`options.put(Option.A, processorOption.getKey() + '=' + processorOptions.getValue())` sets
`-A=key=value` - the javac compiler recognizes it
`options.put("-A" + processorOption.getKey() + "=" + processorOption.getValue(), Boolean.toString(true))` sets
`-A=key=value=true` - the javac compiler doesn't recognize it

Comment (snjeza) [reply org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacUtils.java]: You can use `options.put(Option.A,...` or`options.put("-A",...`
See https://github.com/openjdk/jdk/blob/40ae4699622cca72830acd146b7b5c4efd5a43ec/src/jdk.compiler/share/classes/com/sun/tools/javac/util/Options.java#L149C4-L155C6
```
    public void put(String name, String value) {
        values.put(name, value);
    }


    public void put(Option option, String value) {
        values.put(option.primaryName, value);
    }
```

Comment (snjeza) [reply org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacUtils.java]: The javac compiler uses `Option.IsSet` at
https://github.com/openjdk/jdk/blob/e023addf701ce4321040c96bd501355ece75a05c/src/jdk.compiler/share/classes/com/sun/tools/javac/main/JavaCompiler.java#L1301-L1311
```
    static boolean explicitAnnotationProcessingRequested(Options options, JavaFileManager fileManager) {
        return
            options.isSet(PROCESSOR) ||
            options.isSet(PROCESSOR_PATH) ||
            options.isSet(PROCESSOR_MODULE_PATH) ||
            options.isSet(PROC, "only") ||
            options.isSet(PROC, "full") ||
            options.isSet(A) ||
            options.isSet(XPRINT) ||
            fileManager.hasLocation(ANNOTATION_PROCESSOR_PATH);
        // Skipping -XprintRounds and -XprintProcessorInfo
```

Comment (testforstephen) [reply org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacUtils.java]: > The javac compiler uses `Option.IsSet` at https://github.com/openjdk/jdk/blob/e023addf701ce4321040c96bd501355ece75a05c/src/jdk.compiler/share/classes/com/sun/tools/javac/main/JavaCompiler.java#L1301-L1311
> 
> ```
>     static boolean explicitAnnotationProcessingRequested(Options options, JavaFileManager fileManager) {
>         return
>             options.isSet(PROCESSOR) ||
>             options.isSet(PROCESSOR_PATH) ||
>             options.isSet(PROCESSOR_MODULE_PATH) ||
>             options.isSet(PROC, "only") ||
>             options.isSet(PROC, "full") ||
>             options.isSet(A) ||
>             options.isSet(XPRINT) ||
>             fileManager.hasLocation(ANNOTATION_PROCESSOR_PATH);
>         // Skipping -XprintRounds and -XprintProcessorInfo
> ```

This is not the code that will consume the processor options. The link below is how Javac compiler parses the processor options from the Options object.

https://github.com/openjdk/jdk/blob/e023addf701ce4321040c96bd501355ece75a05c/src/jdk.compiler/share/classes/com/sun/tools/javac/processing/JavacProcessingEnvironment.java#L639

The previous implementation in this PR https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/pull/820 should work well.

Comment (snjeza) [reply org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacUtils.java]: You are right. 
JavacProcessingEnvironment.initProcessorOptions() expects -`Akey=value=true`. 
Options.isSet(Option.A) expects `-A=`
I have updated the PR.

**Comment by snjeza** [comment]
@datho7561 @mickaelistria @testforstephen  I have updated the PR.

**Comment by snjeza** [comment]
test this please

**Comment by datho7561** [comment]
This doesn't work for me anymore when running through Eclipse IDE (I get the original bug you described). I'm running the test suite to see if it works there.

*update*: The test is passing when running through the IDE, but not on command line

**Comment by snjeza** [comment]
> update: The test is passing when running through the IDE, but not on command line

~~I'm working on that.~~

**Comment by snjeza** [comment]
@datho7561 @mickaelistria I have updated the PR. Could you, please, try it now.

**Comment by datho7561** [comment]
I still get the errors:

image

**Comment by snjeza** [comment]
> I still get the errors:

I can't reproduce the issue. Could you, please, try the following:
- rebuild the project
- edit or reopen Main.java

RegressionTests.testlombok() passes - https://ci.eclipse.org/ls/job/jdt-core-incubator/job/PR-992/15/consoleFull
```
01:16:08  Tests run: 6, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 15.03 s -- in org.eclipse.jdt.core.tests.javac.RegressionTests
01:16:08  org.eclipse.jdt.core.tests.javac.RegressionTests.testlombok -- Time elapsed: 7.262 s
01:16:08  org.eclipse.jdt.core.tests.javac.RegressionTests.testBuildMultipleOutputDirectories2 -- Time elapsed: 6.707 s
01:16:08  org.eclipse.jdt.core.tests.javac.RegressionTests.testCheckBuild -- Time elapsed: 0.059 s
01:16:08  org.eclipse.jdt.core.tests.javac.RegressionTests.testBuildMultipleOutputDirectories -- Time elapsed: 0.063 s
01:16:08  org.eclipse.jdt.core.tests.javac.RegressionTests.testGetDOMForClassWithSource -- Time elapsed: 0.540 s
01:16:08  org.eclipse.jdt.core.tests.javac.RegressionTests.testBuildReferenceOtherProjectSource -- Time elapsed: 0.077 s
```
With the PR
lwith

Without the PR
lwithout

**Comment by datho7561** [comment]
:facepalm: I needed to enable annotation in the project settings. After I did it worked flawlessly. Sorry about that

**Review by datho7561 — APPROVED** [review]
Looks good. Thanks, Snjezana!

**Pull request merged**